### ISSUE TITLE
stack frames: a tangle of problems

### ISSUE THREAD

**Issue by matsc-at-sics-se** [matsc-at-sics-se/unison#19]
**Some rules for FrameSize and offset relocation on x86**

- Callee-saved spill slots created with push/pop apparently do not occur in the stack or `fixedStack` sections.  Let `CSSlotSize` denote their total size, used below.
- `FrameSize` should *not* include input args passed on stack *nor* callee-saved spill slots.
- `FrameSize` should be a multiple of 8.
- In the presence of non-tail-recursive calls, `(FrameSize+CSSlotSize+8) mod 16 = 0` must hold.
- `%rsp` increment/decrement instructions should be optional, active iff `FrameSize > 0`.
- Frame indices in the frame proper (not input args passed on stack) should be relocated wrt. `FrameSize`.
- Frame indices for input args passed on stack should be relocated wrt. `(FrameSize+CSSlotSize+8)`.
- Frame indices for input args passed on stack should all be 8-aligned and never subject to packing.
- Suggestion: Should not `fixedStack` be reserved for input args passed on stack? I.e. all spill slots should go to `stack`, not `fixedStack`? It would simplify all of the above.

Here is a piece of code where most of this goes wrong.

**C code**
```
extern void g(int,int,int,int,int,int,int,int,int,int *);
extern void h(int,int,int,int,int,int,int,int,int,int);

void f(int a0,int a1,int a2,int a3,int a4,int a5,int a6,int a7,int a8,int a9) {
  int arr[7];
  g(a1,a2,a3,a4,a5,a6,a7,a8,a9,arr);
  h(a2,a3,a4,a5,a6,a7,a8,a9,a0,a1);
}
```
**Stack Frame Layout**
```
0x0000: | in argument #10|
0x0008: | in argument #9 |
0x0010: | in argument #8 |
0x0018: | in argument #7 |
0x0020: | return address |
0x0028: | spilled callee-saved (optional) |
0x0030: | spilled callee-saved (optional) |
0x0038: | spilled callee-saved (optional) |
0x0040: | spilled callee-saved (optional) |
0x0048: | spilled callee-saved (optional) |
0x0050: | spilled callee-saved (optional) |
0x0058: | alignment pad |
0x0060: | a[6] |    pad |
0x0068: | a[4] |   a[5] |
0x0070: | a[2] |   a[3] |
0x0078: | a[0] |   a[1] |
0x0080: | out argument #10|
0x0088: | out argument #9 |
0x0090: | out argument #8 |
0x0098: | out argument #7 |
0x00a0:
```
**Value of %rsp**

Just before call to f(): `0x0020`
Just after call to f(): `0x0028`
Inside f(): `0x00a0`

**LLVM generated assembly**
```
f:      pushq   %rbp    // spill callee-saved
        pushq   %r15
        pushq   %r14
        pushq   %r13
        pushq   %r12
        pushq   %rbx
        subq    $72, %rsp       // decr %rsp from 0x0058 to 0x00a0
        movl    %r9d, %r14d     // move a5
        movl    %r8d, %r15d     // move a4
        movl    %ecx, %r12d     // move a3
        movl    %edx, %ebp      // move a2
        movl    %esi, %ebx      // move a1
        movl    %edi, %r13d     // spill a0
        movl    128(%rsp), %r9d // move a6
        movl    136(%rsp), %ecx // move a7
        movl    144(%rsp), %edx // move a8
        movl    152(%rsp), %esi // move a9
        leaq    32(%rsp), %rax  // get &arr[0]
        movq    %rax, 24(%rsp)  // put &arr[0] in out #10
        movl    %esi, 16(%rsp)  // put a9 in out #9
        movl    %edx, 8(%rsp)   // put a8 in out #8
        movl    %ecx, (%rsp)    // put a7 in out #7
        movl    %ebx, %edi      // move a1
        movl    %ebp, %esi      // move a2
        movl    %r12d, %edx     // move a3
        movl    %r15d, %ecx     // move a4
        movl    %r14d, %r8d     // move a5
        callq   g
        movl    %ebx, 24(%rsp)  // put a1 in out #10
        movl    %r13d, 16(%rsp) // unspill a0 in out #9
        movl    152(%rsp), %eax // get a9
        movl    %eax, 8(%rsp)   // put a9 in out #8
        movl    144(%rsp), %eax // get a8
        movl    %eax, (%rsp)    // put a8 in out #7
        movl    %ebp, %edi      // move a2
        movl    %r12d, %esi     // move a3
        movl    %r15d, %edx     // move a4
        movl    %r14d, %ecx     // move a5
        movl    128(%rsp), %r8d // move a6
        movl    136(%rsp), %r9d // move a7
        callq   h
        addq    $72, %rsp       // incr %rsp from 0x00a0 to 0x0058
        popq    %rbx    // unspill callee-saved
        popq    %r12
        popq    %r13
        popq    %r14
        popq    %r15
        popq    %rbp
        retq
```
**Unison generated assembly**

Note: code to build output args #7 to #10 is scheduled too late: some dependency is missing.

Instructions below are annotated with _intended_ meaning.  Pretty much all offsets are wrong.
```
name:            f
fixedStack:      
  - { id: 0, offset: -28, size: 4, alignment: 8 }       // in arg #10
  - { id: 1, offset: -20, size: 4, alignment: 16 }      // in arg #9
  - { id: 2, offset: -12, size: 4, alignment: 8 }       // in arg #8
  - { id: 3, offset: -4, size: 4, alignment: 16 }       // in arg #7
  - { id: 4, offset: -8, size: 4, alignment: 4 }        // spill slot for a2
  - { id: 5, offset: -16, size: 4, alignment: 4 }       // spill slot for a3
  - { id: 6, offset: -24, size: 4, alignment: 4 }       // spill slot for a4
  - { id: 7, offset: -32, size: 4, alignment: 4 }       // spill slot for a5
  - { id: 8, offset: -36, size: 4, alignment: 4 }       // spill slot for a6 (silly)
stack:           
  - { id: 0, offset: -76, size: 28, alignment: 16 }     // arr[]
body:             |
  bb.0 (freq 1):
  
    %rsp = SUB64ri32 %rsp, 76               // allocate frame
    MOV32mr %rsp, 1, _, 68, _, %edx         // store a2 at id: 4
    MOV32mr %rsp, 1, _, 60, _, %ecx         // store a3 at id: 5
    %edx = MOV32rm %rsp, 1, _, 60, _        // load a3 at id: 5
    MOV32mr %rsp, 1, _, 52, _, %r8d         // store a4 at id: 6
    %ecx = MOV32rm %rsp, 1, _, 52, _        // load a4 at id: 6
    MOV32mr %rsp, 1, _, 44, _, %r9d         // store a5 at id: 7
    %r8d = MOV32rm %rsp, 1, _, 44, _        // load a5 at id: 7
    %r9d = MOV32rm %rsp, 1, _, 72, _        // load a6 at id: 3 (in arg #7)
    MOV32mr %rsp, 1, _, 40, _, %r9d         // store a6 at id: 8
    MOV32mr %rsp, 1, _, 16, _, %edi         // store a0 at out #9 for h()
    %edi = MOV32rr %esi                     // move a1
    MOV32mr %rsp, 1, _, 24, _, %esi         // store a1 at out #10 for h()
    %esi = MOV32rm %rsp, 1, _, 68, _        // load a2 at id: 4
    CALL64pcrel32 @g                        // call g(a1,a2,a3,a4,a5,a6,_,_,a0,a1)
    %r9d = MOV32rm %rsp, 1, _, 64, _        // load a7 at id: 2 (in arg #8)
    MOV32mr %rsp, 1, _, 0, _, %r9d          // store a7 at out #7 for g()
    %r8d = MOV32rm %rsp, 1, _, 40, _        // load a6 at id: 8
    %edi = MOV32rm %rsp, 1, _, 68, _        // load a2 at id: 4
    %esi = MOV32rm %rsp, 1, _, 60, _        // load a3 at id: 5
    %edx = MOV32rm %rsp, 1, _, 52, _        // load a4 at id: 6
    %ecx = MOV32rm %rsp, 1, _, 44, _        // load a5 at id: 7
    CALL64pcrel32 @h                        // call h(a2,a3,a4,a5,a6,a7,a7,_,a0,a1)
    %rsp = ADD64ri32 %rsp, 76               // deallocate frame
    %eax = MOV32rm %rsp, 1, _, 56, _        // load a8 at id: 1 (in arg #9)
    MOV32mr %rsp, 1, _, 8, _, %eax          // store a8 at out #8 for g()
    MOV32mr %rsp, 1, _, 0, _, %eax          // store a8 at out #7 for h()
    %eax = MOV32rm %rsp, 1, _, 48, _        // load a9 at id: 0 (in arg #10)
    MOV32mr %rsp, 1, _, 16, _, %eax         // store a9 at out #9 for g()
    MOV32mr %rsp, 1, _, 8, _, %eax          // store a9 at out #8 for h()
    %rax = LEA64r %rsp, 1, _, 0, _          // load &arr[0]
    MOV64mr %rsp, 1, _, 24, _, %rax         // store &arr[0] at out #10 for g()
    RETQ 42
```

**Comment by matsc-at-sics-se** [comment]
Instruction scheduling bug was due to missing memory read/write side-effects.
Fixed by commit `93761ad54069f3f9bcb8bd77a01d8f7d4369e5e4` and  `a741b80b46444bda73d97d6fa4ec0a78b306f965`.
The other problems are still there.

**Comment by matsc-at-sics-se** [comment]
I had a quick look at the other targets. They are all rather similar:
- The stack grows downwards
- The stack pointer is the base pointer for stack frame accesses
- The "extended" stack frame consists of, by decreasing address:
  - input args beyond the 6th or so
  - return address (X86 only)
  - spilled callee-saved
  - struct & spill slots
- Prologue for Mips/Hexagon: alloca by `%FrameSize`, then spill callee-saved
- Prologue for ARM/X86: push callee-saved, then alloca by `%FrameSize`

**Comment by matsc-at-sics-se** [comment]
The recent commits up to and including `8d081b41cabcc2f57c4bf1ba6bcdced86173a246` seem to fix most of it.  One problem remains: function calls with >6 arguments, which are passed on the stack.  No space is currently allocated for that.  For example, *.c:

```
extern void foo(char *a, short int *b, int *c, long *d, int, int, int, int);
extern void bar(void);
void frame(int a, int b, int c, int d, int e, int f,
	   long g, int h, short i, char j) {
  long fd[2];
  int fc[3];
  short fb[5];
  char fa[9];
  bar();
  fa[0] = j;
  fb[0] = i;
  fc[0] = h;
  fd[0] = g;
  foo(fa, fb, fc, fd, 7, 7, 7, 7);
}
```
*.uni:
```
// @lint: --nomustconflicts=false --nocostoverflow=false --nocomponentconflicts=false
function: frame
b0 (entry, exit, return, freq: 1):
    o0: [t0:rbx,t1:rbp,t2:r12,t3:r13,t4:r14,t5:r15,t6:rsp,t7:rip] <- (in) []
    o1: [t8] <- MOV64rm [%fixed-stack.3,1,_,0,_] (mem: 0)
    o2: [t9] <- MOV32rm [%fixed-stack.2,1,_,0,_] (mem: 0)
    o3: [t10] <- MOV16rm [%fixed-stack.1,1,_,0,_] (mem: 0)
    o4: [t11] <- MOV8rm [%fixed-stack.0,1,_,0,_] (mem: 0)
    o5: [] <- CALL64pcrel32 [@bar]
    o6: [t12:rax,t13:rcx_rdx,t14:rsi_rdi,t15:r8_r9,t16:r10_r11,t17:ymm0,t18:ymm1_15] <- (fun) [] (call: o5)
    o7: [] <- (kill) [t12,t13,t14,t15,t16,t17,t18]
    o8: [] <- MOV8mr [%stack.3,1,_,0,_,t11] (mem: 0)
    o9: [] <- MOV16mr [%stack.2,1,_,0,_,t10] (mem: 0)
    o10: [] <- MOV32mr [%stack.1,1,_,0,_,t9] (mem: 0)
    o11: [] <- MOV64mr [%stack.0,1,_,0,_,t8] (mem: 0)
    o12: [] <- MOV32mi [t6,1,_,8,_,7] (mem: 0)
    o13: [] <- MOV32mi [t6,1,_,0,_,7] (mem: 0)
    o14: [t19] <- LEA64r [%stack.3,1,_,0,_]
    o15: [t20] <- LEA64r [%stack.2,1,_,0,_]
    o16: [t21] <- LEA64r [%stack.1,1,_,0,_]
    o17: [t22] <- LEA64r [%stack.0,1,_,0,_]
    o18: [t23] <- MOV32ri [7]
    o19: [t24] <- (copy) [t23]
    o20: [] <- CALL64pcrel32 [@foo]
    o21: [t25:rax,t26:rcx_rdx,t27:rsi_rdi,t28:r8_r9,t29:r10_r11,t30:ymm0,t31:ymm1_15] <- (fun) [t19:rdi,t20:rsi,t21:rdx,t22:rcx,t23:r8d,t24:r9d] (call: o20)
    o22: [] <- (kill) [t25,t26,t27,t28,t29,t30,t31]
    o23: [] <- RETQ [42]
    o24: [] <- (out) [t0:rbx,t1:rbp,t2:r12,t3:r13,t4:r14,t5:r15,t6:rsp,t7:rip]
```
So the only hint that `foo` takes 8 arguments is `o12-o13`. Similarly for *.alt.uni. In theory, one could parse for explicit writes to the stack. That strikes me as awkward and brittle.

Note that *.uni and *.alt.uni also contain, in other sections:
```
define void @frame(i32 %a, i32 %b, i32 %c, i32 %d, i32 %e, i32 %f, i64 %g, i32 %h, i16 signext %i, i8 signext %j) #0 {
...
  call void @foo(i8* %4, i16* %5, i32* %6, i64* %7, i32 7, i32 7, i32 7, i32 7) #3
...
}
```
```
declare void @foo(i8*, i16*, i32*, i64*, i32, i32, i32, i32) #2
```
So how can I best compute: **the max arity of the functions called by the function being compiled**?

**Comment by robcasloz** [comment]
Can you post here the .mir and .asm.mir files for the `frame` function?

**Comment by matsc-at-sics-se** [comment]
Attached is that version.
By the way, for computing the stack frame size, I would like to know whether or not the function contains any non-tail-recursive calls. Can that be checked in constant time?
[frame.tgz.txt](https://github.com/matsc-at-sics-se/unison/files/1973698/frame.tgz.txt)

**Comment by matsc-at-sics-se** [comment]
Check for non-tail-recursive call implemented in commit `176748af7bca4d495380bace138ad656a769cd13`. In linear time, though.

**Comment by matsc-at-sics-se** [comment]
Commit `0b28cb4af4ed0d9194aa96365a08214538353f81` adds the `ExportPreLow` transformation `liftStackArgSize`. That closes the issue.
